src/dispatcher.rs

```diff
@@ -0,0 +1,162 @@
+use std::sync::Arc;
+
+#[cfg(feature = "prefetch")]
+use ethcore::transaction::Action;
+use ethcore::transaction::SignedTransaction;
+use failure::{Error, Fail, Fallible, ResultExt};
+use serde_bytes::ByteBuf;
+
+use oasis_core_keymanager_client::KeyManagerClient;
+#[cfg(feature = "prefetch")]
+use oasis_core_runtime::storage::mkvs::Prefix;
+use oasis_core_runtime::{
+    common::{cbor, crypto::hash::Hash, roothash::Message as RoothashMessage},
+    transaction::{
+        context::Context,
+        dispatcher::{CheckOnlySuccess, Dispatcher as TxnDispatcher},
+        tags::Tags,
+        types::*,
+    },
+};
+
+use super::{
+    block::OasisBatchHandler,
+    methods::{check, execute},
+};
+
+use oasis_runtime_api as api;
+
+/// Dispatch error.
+#[derive(Debug, Fail)]
+enum DispatchError {
+    #[fail(display = "method not found: {}", method)]
+    MethodNotFound { method: String },
+}
+
+pub struct DecodedCall {
+    pub transaction: SignedTransaction,
+}
+
+pub struct Dispatcher {
+    /// Registered batch handler.
+    batch_handler: Box<OasisBatchHandler>,
+}
+
+impl Dispatcher {
+    /// Create a new runtime method dispatcher instance.
+    pub fn new(key_manager: Arc<dyn KeyManagerClient>) -> Dispatcher {
+        Dispatcher {
+            batch_handler: Box::new(OasisBatchHandler::new(key_manager)),
+        }
+    }
+
+    fn decode_transaction(&self, call: &Vec<u8>, ctx: &mut Context) -> Fallible<DecodedCall> {
+        let call: TxnCall = cbor::from_slice(call).context("unable to parse call")?;
+
+        if call.method != api::METHOD_TX {
+            return Err(DispatchError::MethodNotFound {
+                method: call.method,
+            }
+            .into());
+        }
+
+        let call_args: ByteBuf =
+            cbor::from_value(call.args).context("unable to parse call arguments")?;
+        let signed_transaction = check::tx(&call_args, ctx)?;
+
+        Ok(DecodedCall {
+            transaction: signed_transaction,
+        })
+    }
+
+    fn encode_response(&self, call: &DecodedCall, ctx: &mut Context) -> Fallible<Vec<u8>> {
+        let response = execute::tx(call, ctx)?;
+        let response = TxnOutput::Success(cbor::to_value(response));
+        Ok(cbor::to_vec(&response))
+    }
+
+    fn map_error(&self, err: &Error) -> Vec<u8> {
```

Comment:
hmm, I'm not quite sure this is really a `map` operation. Maybe more of a `fn serialize_check_result` or something